core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java

```diff
@@ -116,20 +118,17 @@ private UnderFileSystem getOrAdd(AlluxioURI ufsUri, UnderFileSystemConfiguration
     if (cachedFs != null) {
       return cachedFs;
     }
-    UnderFileSystem fs = UnderFileSystem.Factory.create(ufsUri.toString(), ufsConf);
-    cachedFs = mUnderFileSystemMap.putIfAbsent(key, fs);
-    if (cachedFs == null) {
-      // above insert is successful
+    // On cache miss, synchronize the creation to ensure ufs is only created once
+    synchronized (mLock) {
+      cachedFs = mUnderFileSystemMap.get(key);
+      if (cachedFs != null) {
+        return cachedFs;
+      }
+      UnderFileSystem fs = UnderFileSystem.Factory.create(ufsUri.toString(), ufsConf);
+      mUnderFileSystemMap.putIfAbsent(key, fs);
```

Comment:
Should we check that this returns null?

Reply:
it is not expected to be null due to the synchronization